Align value module ordering with OpenCypher

fickludd opened the pull request:
This PR changes the ordering of storable and virtual `Value`s. This is done to align with the accepted [OpenCypher CIP defining orderability](https://github.com/opencypher/openCypher/blob/master/cip/1.accepted/CIP2016-06-14-Define-comparability-and-equality-as-well-as-orderability-and-equivalence.adoc).

Changes to value ordering when comparing **storable** values:
  * Arrays are ordered before scalars

Changes to value ordering when comparing **all** values:
  * supertypes are ordered as `map`, `node`, `relationship`, `list/array`, `path`, `point`, `text`, `boolean`, `number`, `NO_VALUE`
  * lists have dictionary order, eg. `[], [1], [1,2], [2], ...` 
  * paths are ordered as lists of alternating nodes and relationships

Finally this allows us to use `AnyValues.COMPARATOR` for `SortPipe` ordering.

fickludd committed:
Change Value order to align with the Cypher standard.

Not that this changes the order between the storable values. It also changes arrays to dictionary sort order, instead of being sorted by length first.

fickludd committed:
Co-order ListValue and ArrayValue, as specified by OpenCypher

fickludd committed:
Order PathValue as specified by OpenCypher

fickludd committed:
Change SortPipe to use AnyValues.COMPARATOR for ordering

pontusmelke commented on community/values/src/main/java/org/neo4j/values/SequenceValue.java:
```diff
@@ -52,4 +55,25 @@ default boolean equals( SequenceValue other )
+
+        while ( x == 0 && i < length )
+        {
+            x = comparator.compare( value( i ), other.value( i ) );
```
I know this doesn't really change from how it used to work but not all ListValue implementations have constant time random access - `FilteredListValue` hasn't. So it would be more efficient - at least in that case - to use the `iterator` instead.

fickludd replied:
Nice catch. Specialized both `compareTo` and `equals` now.

fickludd committed:
Specialize SequenceValue equals and compareTo

Some SequenceValue implementations have list like behaviour, and do thus
not support constant speed random access. In these cases it will be much
faster to iterate over the sequence using an Iterator.

We introduce a way to discover the preferred method of iteration from any
sequence value, and specialize the SequenceValue equals and compareTo so
that they are always O(n+m), where n and m are the sizes of the
sequences.

fickludd commented:
@tinwelint For future reference: should we use the `team-kernel` label instead of `kernel` or what is happening here?

tinwelint commented:
There's only one label now: `team-kernel`
